blockdag/reachability.go

```diff
@@ -366,14 +416,199 @@ func (rtn *reachabilityTreeNode) propagateInterval(subTreeSizeMap map[*reachabil
 				child.setInterval(childInterval)
 				queue = append(queue, child)
 			}
+		}
+
+		modifiedTreeNodes[current] = struct{}{}
+	}
+	return modifiedTreeNodes, nil
+}
+
+func (rtn *reachabilityTreeNode) reindexIntervalsEarlierThanReindexRoot(
+	reindexRoot *reachabilityTreeNode) (map[*reachabilityTreeNode]struct{}, error) {
+
+	commonAncestor := rtn.findCommonAncestor(reindexRoot)
+	commonAncestorChosenChild, err := commonAncestor.findReachabilityTreeAncestorInChildren(rtn)
+	if err != nil {
+		return nil, err
+	}
+
+	if rtn.interval.end < commonAncestorChosenChild.interval.start {
+		// rtn is in the subtree before the chosen child
+		return rtn.reclaimIntervalBeforeChosenChild(commonAncestor, commonAncestorChosenChild, reindexRoot)
+	}
+	if commonAncestorChosenChild.interval.end < rtn.interval.start {
+		// rtn is in the subtree after the chosen child
+		return rtn.reclaimIntervalAfterChosenChild(commonAncestor, commonAncestorChosenChild, reindexRoot)
+	}
+	return nil, errors.Errorf("rtn is in the chosen child's subtree")
+}
+
+func (rtn *reachabilityTreeNode) reclaimIntervalBeforeChosenChild(
+	commonAncestor *reachabilityTreeNode, commonAncestorChosenChild *reachabilityTreeNode, reindexRoot *reachabilityTreeNode) (
+	map[*reachabilityTreeNode]struct{}, error) {
+
+	modifiedTreeNodes := make(map[*reachabilityTreeNode]struct{})
+
+	current := commonAncestorChosenChild
+	for !current.hasSlackIntervalBefore() {
+		if current == reindexRoot {
```

Comment:
The code here is not linear. Anything happening inside this if clause actually happens in the end of the loop.

Therefore I suggest to either move this whole thing into a function, or break out of here and do this outside the loop